### ISSUE TITLE
Add localization support

### ISSUE THREAD

**Issue by alnlarsen** [opentap/opentap#1696]
Find a general way to allow OpenTAP plugin developers to support multiple languages.

The primary features where this needs to be supported would be: 

1. Display attributes (name, description, etc)
2. Log messages
3. (potentially) CLI actions, though this gets weird in e.g. a scripting environment. Maybe the language only applies to descriptions, and different languages are aliases for the same option.

The user needs a way to select a preferred language (maybe a ranked list), and plugins need to declare what languages they can support.

Let's look into how this is handled in other frameworks, and discuss how this can be implemented in OpenTAP in a backwards compatible way.

**Comment by alnlarsen** [comment]
We also discussed an alternative to language packs: Adding an optional `culture` argument to the existing `[Display]` attribute, and then allowing multiple display attributes

```csharp
[Display("English Name", "English Description")]
[Display("German Name", "German Description", language: "de")]
public class MyTestStep : TestStep { /* ... */ }
```

This is would by far be the easiest way for plugins to provide support multiple languages. A (non-exhaustive) list of problems with this approach:
1. It places the responsibility for providing translations on the plugin author(s)
2. Translations must be directly embedded in the source code
3. User-provided translations of non- open source plugins is not possible without contacting the authors
4. Authors of open-source plugins are not necessarily able to evaluate the quality of translations suggested through pull requests
5. It adds a lot of noise to the source code. (Imagine scenarios with support for 10 languages)

Even with these drawbacks, I think it is worth considering. I think this system can be neatly integrated with the proposed language-pack system.

**Comment by alnlarsen** [comment]
I think we can improve localization support gradually. I would propose splitting this issue into the following tasks:

1. OpenTAP should add a language selection component setting,  `LocalizationSettings`
2. Add support for localized display attributes. This attribute may initially be chosen based on the `language` property, but in the long term, OpenTAP can inject translations in the Annotation layer.
3. Add support for language packs. (e.g. a user-provided mapping of translations). We can decide the format of a language pack, and the context clues we should support when we implement this, but regardless of the specifics of the implementation, the output should be a display attribute with a language.
4. Add support for localizing log messages. When 1) is implemented, plugin authors can manually localize logs by checking `LocalizationSettings.Current.Language`, but in the future, we may want to provide additional localization features specifically for logging.

**Comment by timremm** [comment]
Something that may relate with this, is the handling of different decimal separator. 
English: 10,000.123
German (and maybe other languages: 10.000,123
When I experiment with localization on my own, I changed the way how the separator are being treated. 
Eventually this can be supported here as well (either as a separate setting, or when selecting German / English it adjusted automatically)

**Comment by rmadsen-ks** [comment]
> Something that may relate with this, is the handling of different decimal separator. English: 10,000.123 German (and maybe other languages: 10.000,123 When I experiment with localization on my own, I changed the way how the separator are being treated. Eventually this can be supported here as well (either as a separate setting, or when selecting German / English it adjusted automatically)

@timremm, What about stored data vs displayed data? For example the CSV result listener might need to store in a format that is machine readable, but since cultures may change across stations, this might be difficult to support. What do you think about this aspect?

**Comment by timremm** [comment]
> > Something that may relate with this, is the handling of different decimal separator. English: 10,000.123 German (and maybe other languages: 10.000,123 When I experiment with localization on my own, I changed the way how the separator are being treated. Eventually this can be supported here as well (either as a separate setting, or when selecting German / English it adjusted automatically)
> 
> [@timremm](https://github.com/timremm), What about stored data vs displayed data? For example the CSV result listener might need to store in a format that is machine readable, but since cultures may change across stations, this might be difficult to support. What do you think about this aspect?

Not sure if this cover all use cases, but the Result class and Metric Manager could convert it to the standard representation. But not sure how important it is to support this.

**Comment by rmadsen-ks** [comment]
> Not sure if this cover all use cases, but the Result class and Metric Manager could convert it to the standard representation. But not sure how important it is to support this.

The way I see it, this should only affect how its displayed in the UI. Maybe there could be a setting specific to the result listener when needed.

**Comment by timremm** [comment]
> > Not sure if this cover all use cases, but the Result class and Metric Manager could convert it to the standard representation. But not sure how important it is to support this.
> 
> The way I see it, this should only affect how information is displayed in the UI. Maybe there could be a setting specific to the result listener when needed.

Yes indeed. If the stored data should support this, there should be settings like "decimal separator" and "thousand separator"

**Comment by alnlarsen** [comment]
A few additional points after another round of discussion:

* We don't want to go for the multiple display attribute approach. We don't want to implement and support multiple different localization systems, and due to the drawbacks listed above we don't find it flexible enough (although the simplicity is tempting)
* We will support "language packs", i.e. translations provided by a file (maybe XML, exact format TBD)
* We should create tooling for generating a comprehensive translation template (`tap translation create <my_dll>`). We can use reflection to generate an exhaustive template without access to the source code.
* Provide tooling for updating translations (only add missing entries to existing file, create new file, merge files, TBD)
